scripts/esm/xr-navigation.mjs

```diff
@@ -0,0 +1,123 @@
+
+    /** @type {Map<XrInputSource, boolean>} */
+    activePointers = new Map();
+
+    validColor = new Color(0, 1, 0);    // Green for valid teleport
+
+    invalidColor = new Color(1, 0, 0);   // Red for invalid teleport
+
+    /** @type {Map<XrInputSource, { handleSelectStart: Function, handleSelectEnd: Function }>} */
+    inputHandlers = new Map();
+
+    initialize() {
+        this.app.xr.input.on('add', (inputSource) => {
+            const handleSelectStart = () => {
+                this.activePointers.set(inputSource, true);
+            };
+
+            const handleSelectEnd = () => {
+                this.activePointers.set(inputSource, false);
+                this.tryTeleport(inputSource);
+            };
+
+            // Attach the handlers
+            inputSource.on('selectstart', handleSelectStart);
+            inputSource.on('selectend', handleSelectEnd);
+
+            // Store the handlers in the map
+            this.inputHandlers.set(inputSource, { handleSelectStart, handleSelectEnd });
+            this.inputSources.add(inputSource);
+        });
+
+        this.app.xr.input.on('remove', (inputSource) => {
+            const handlers = this.inputHandlers.get(inputSource);
+            if (handlers) {
+                inputSource.off('selectstart', handlers.handleSelectStart);
+                inputSource.off('selectend', handlers.handleSelectEnd);
+                this.inputHandlers.delete(inputSource);
+            }
+            this.activePointers.delete(inputSource);
+            this.inputSources.delete(inputSource);
+        });
+    }
+
+    findPlaneIntersection(origin, direction) {
+        // Find intersection with y=0 plane
+        if (Math.abs(direction.y) < 0.00001) return null;  // Ray is parallel to plane
+
+        const t = -origin.y / direction.y;
+        if (t < 0) return null;  // Intersection is behind the ray
+
+        return new Vec3(
+            origin.x + direction.x * t,
+            0,
+            origin.z + direction.z * t
+        );
+    }
+
+    tryTeleport(inputSource) {
+        const origin = inputSource.getOrigin();
+        const direction = inputSource.getDirection();
+
+        const hitPoint = this.findPlaneIntersection(origin, direction);
+        if (hitPoint) {
+            const cameraY = this.entity.getPosition().y;
+            hitPoint.y = cameraY;
+            this.entity.setPosition(hitPoint);
+        }
+    }
+
+    update() {
+        for (const inputSource of this.inputSources) {
+            // Only show ray when trigger is pressed
+            if (!this.activePointers.get(inputSource)) continue;
+
+            const start = inputSource.getOrigin();
+            const direction = inputSource.getDirection();
+
+            const hitPoint = this.findPlaneIntersection(start, direction);
+
+            if (hitPoint) {
+                // Draw line to intersection point
+                this.app.drawLine(start, hitPoint, this.validColor);
```

Comment:
you could move all the drawing code to a separate function, allowing users to easily extend the script and implement a custom drawing by overriding the function

Reply:
Yes, customizing visuals is important but I'd like to leave solving that to a follow up PR.